File: packages/node/lib/services/terminal.ts
import * as grpc from '@grpc/grpc-js';
import {
  ConnectToProcessRequest,
  ListProcessesRequest,
  ListProcessesResponse,
  ProcessEvent,
  SendInputRequest,
  SendInputResponse,
  SendSignalRequest,
  SendSignalResponse,
  SpawnProcessRequest,
  TerminalClient,
  UpdateProcessRequest,
  UpdateProcessResponse
} from '../generated/terminal';
import { METALLIC_AGENT_PORT } from '../utils/constants';

export class TerminalService {
  private client: TerminalClient;

  constructor(projectId: string, instanceId: string) {
    this.client = new TerminalClient(
      `${projectId}-${instanceId}-${METALLIC_AGENT_PORT}.metallic.computer:443`,
      grpc.credentials.createSsl()
    );
  }

  async listProcesses(req: ListProcessesRequest): Promise<ListProcessesResponse> {
    return new Promise((resolve, reject) => {
      this.client.listProcesses(req, (err, response) => {
        if (err) {
          reject(err);
          return;
        }

        if (!response.success) {
          reject(new Error(response.error));
          return;
        }

        resolve(response);
      });
    });
  }

  spawnProcess(req: SpawnProcessRequest): grpc.ClientReadableStream<ProcessEvent> {
    return this.client.spawnProcess(req);
  }

  connectToProcess(req: ConnectToProcessRequest): grpc.ClientReadableStream<ProcessEvent> {
    return this.client.connectToProcess(req);
  }

  async sendInput(req: SendInputRequest): Promise<SendInputResponse> {
    return new Promise((resolve, reject) => {
      this.client.sendInput(req, (err, response) => {
        if (err) {
          reject(err);
          return;
        }

        if (!response.success) {
          reject(new Error(response.error));
          return;
        }

        resolve(response);
      });
    });
  }

  async updateProcess(req: UpdateProcessRequest): Promise<UpdateProcessResponse> {
    return new Promise((resolve, reject) => {
      this.client.updateProcess(req, (err, response) => {
        if (err) {
          reject(err);
          return;
        }

        if (!response.success) {
          reject(new Error(response.error));
          return;
        }

        resolve(response);
      });
    });
  }

  async sendSignal(req: SendSignalRequest): Promise<SendSignalResponse> {
    return new Promise((resolve, reject) => {
      this.client.sendSignal(req, (err, response) => {
        if (err) {
          reject(err);
          return;
        }

        if (!response.success) {
          reject(new Error(response.error));
          return;
        }

        resolve(response);
      });
    });
  }
}

File: packages/python/metallic/services/terminal.py
"""Terminal gRPC service client."""
import grpc
from typing import AsyncIterator

from ..generated import terminal_pb2, terminal_pb2_grpc


class TerminalService:
    """gRPC client for Terminal service."""
    
    def __init__(self, project_id: str, instance_id: str, host: str = "localhost", port: int = 50051):
        """Initialize the Terminal service client.
        
        Args:
            project_id: Project ID
            instance_id: Instance ID
            host: gRPC server host
            port: gRPC server port
        """
        self.project_id = project_id
        self.instance_id = instance_id
        self.host = host
        self.port = port
        self.channel = grpc.aio.insecure_channel(f"{host}:{port}")
        self.stub = terminal_pb2_grpc.TerminalStub(self.channel)
    
    async def close(self):
        """Close the gRPC channel."""
        await self.channel.close()
    
    async def list_processes(self, request: terminal_pb2.ListProcessesRequest) -> terminal_pb2.ListProcessesResponse:
        """List running processes."""
        return await self.stub.ListProcesses(request)
    
    async def spawn_process(self, request: terminal_pb2.SpawnProcessRequest) -> AsyncIterator[terminal_pb2.ProcessEvent]:
        """Spawn a new process."""
        async for event in self.stub.SpawnProcess(request):
            yield event
    
    async def connect_to_process(self, request: terminal_pb2.ConnectToProcessRequest) -> AsyncIterator[terminal_pb2.ProcessEvent]:
        """Connect to an existing process."""
        async for event in self.stub.ConnectToProcess(request):
            yield event
    
    async def send_input(self, request: terminal_pb2.SendInputRequest) -> terminal_pb2.SendInputResponse:
        """Send input to a process."""
        return await self.stub.SendInput(request)
    
    async def update_process(self, request: terminal_pb2.UpdateProcessRequest) -> terminal_pb2.UpdateProcessResponse:
        """Update process settings."""
        return await self.stub.UpdateProcess(request)
    
    async def send_signal(self, request: terminal_pb2.SendSignalRequest) -> terminal_pb2.SendSignalResponse:
        """Send signal to a process."""
        return await self.stub.SendSignal(request)
